File: domain/src/main/java/ru/spliterash/spcore/domain/usecase/abstractCases/AbstractFindEntityByIdUseCase.java
package ru.spliterash.spcore.domain.usecase.abstractCases;

import lombok.RequiredArgsConstructor;
import ru.spliterash.spcore.domain.entity.BaseEntity;
import ru.spliterash.spcore.domain.exceptions.DomainException;
import ru.spliterash.spcore.domain.repo.CrudRepository;
import ru.spliterash.spcore.domain.usecase.UseCase;

@RequiredArgsConstructor
public abstract class AbstractFindEntityByIdUseCase<E extends BaseEntity> implements UseCase<String, E> {
    protected final CrudRepository<E, ?> repository;

    @Override
    public E execute(String id) {
        return repository.findById(id).orElseThrow(() -> getNotFoundException(id));
    }

    protected abstract DomainException getNotFoundException(String id);
}

File: domain/src/main/java/ru/spliterash/spcore/domain/usecase/nonBlockingCases/AbstractFindEntityByIdUseCase.kt
package ru.spliterash.spcore.domain.usecase.nonBlockingCases

import lombok.RequiredArgsConstructor
import ru.spliterash.spcore.domain.entity.BaseEntity
import ru.spliterash.spcore.domain.exceptions.DomainException
import ru.spliterash.spcore.domain.repo.CrudRepository
import ru.spliterash.spcore.domain.usecase.NonBlockingUseCase

@RequiredArgsConstructor
abstract class AbstractFindEntityByIdUseCase<E : BaseEntity?>(
    protected val repository: CrudRepository<E, *>
) : NonBlockingUseCase<String, E> {

    override suspend fun execute(id: String): E {
        return repository.findById(id) ?: throw getNotFoundException(id)
    }

    protected abstract fun getNotFoundException(id: String): DomainException
}